backend/inspirehep/pidstore/providers/texkey.py

```diff
@@ -0,0 +1,149 @@
+        used_texkeys = cls.get_all_texkeys_starting_with(texkey)
+        for _ in range(retry_count):
+            random_part = "".join(
+                random.choices(string.ascii_lowercase + string.digits, k=size)
+            )
+            pid_value = f"{texkey}{random_part}"
+            if pid_value not in used_texkeys:
+                return pid_value
+        raise CannotGenerateUniqueTexKey
+
+    @classmethod
+    def get_texkey(cls, texkey):
+        texkey = (
+            PersistentIdentifier.query.with_for_update()
+            .filter(PersistentIdentifier.pid_value == texkey)
+            .filter(PersistentIdentifier.pid_type == cls.pid_type)
+            .one_or_none()
+        )
+        return texkey
+
+    @classmethod
+    def get_all_texkeys_starting_with(cls, texkey):
+        texkeys = (
+            PersistentIdentifier.query.with_entities(PersistentIdentifier.pid_value)
+            .filter(PersistentIdentifier.pid_type == cls.pid_type)
+            .filter(PersistentIdentifier.pid_value.startswith(texkey))
+            .all()
+        )
+        texkeys = set(flatten_list(texkeys))
+        return texkeys
+
+    @classmethod
+    def is_texkey_assigned_correctly(cls, texkey, uuid):
+        texkey = cls.get_texkey(texkey)
+        if (
+            texkey
+            and texkey.object_uuid == uuid
+            and texkey.status == PIDStatus.REGISTERED
+        ):
+            return True
+        elif texkey:
+            raise PIDAlreadyExistsError(texkey.pid_type, texkey.pid_value)
+        return False
+
+    @classmethod
+    def create(
+        cls, object_type=None, object_uuid=None, data=None, pid_value=None, **kwargs
+    ):
+        first_part = cls.get_texkey_first_part(data)
+        second_part = cls.get_texkey_second_part(data)
+        if first_part and second_part:
+            new_texkey = f"{first_part}:{second_part}"
+            pid_value = pid_value or get_value(data, "texkeys[0]")
+            if pid_value and pid_value.startswith(new_texkey):
+                if cls.is_texkey_assigned_correctly(pid_value, object_uuid):
+                    return
+                return super().create(
+                    pid_value=pid_value,
+                    status=PIDStatus.REGISTERED,
+                    object_type=object_type,
+                    object_uuid=object_uuid,
+                    **kwargs,
+                )
+            else:
+                pid_value = cls.get_texkey_with_random_part(new_texkey)
+                if pid_value:
+                    return super().create(
+                        pid_value=pid_value,
+                        status=PIDStatus.REGISTERED,
+                        object_type=object_type,
+                        object_uuid=object_uuid,
+                        **kwargs,
+                    )
+        raise NoAvailableTexKeyFound(object_uuid)
+
+    @classmethod
+    def get(cls, pid_value, object_uuid=None):
```

Comment:
It's there but it works a little different, It's not querying for object_uuid, so it can get pid even if it's assigned to other record. (it gets any PID which pid_value and pid_type and pid_provider are correct). I'm aware that we have primary key on (pid_type, pid_value) but here I want to be sure that requested pid is assigned to this record, and it's faster than doing `if object_uuid == requested_object_uuid)`